Added support for elevation based dark theme overlay color in the Material widget

darrenaustin opened the pull request:
## Description

In order to support the [Material Dark theme specification](https://material.io/design/color/dark-theme.html#properties), this PR adds support for lightening the background color of a `Material` widget when it is in a dark theme based on its elevation.

![mio-design_assets_1A1Uf3a9TsbtgSKT-hjIAaEiEIXFDTUfE_darktheme-elevationsystem](https://user-images.githubusercontent.com/19588/60552610-53910400-9ce4-11e9-880d-33c221bf8204.png)

If the widget is placed in a dark theme (i.e. `Theme.of(context).brightness == Brightness.dark`) a semi-transparent white will be composited over the `Material.color` for the background of the widget. The level of the transparency follows the values in the specification.

If this feature is not wanted, it can be turned off by the new `darkThemeOverlay` boolean property. By default this is turned on, which may lead to some visual changes to existing apps.

## Related Issues

## Tests

I added a group of three new tests to test these features.

## Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [x] Yes, this is a breaking change (Please read [Handling breaking changes]). *Link to announcement email TBD*

<!-- Links -->
[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

darrenaustin committed:
Added support for dark theme overlay colors to the `Material` widget. Also
added a `darkThemeOverlay` boolean to turn it off in cases where it not
wanted.

a-siva committed:
Use the new service protocol message names (#35482)

* Use the new service protocol message names
  clearVMTimeline
  setVMTimelineFlags
  getVMTimeline
  getVMTimelineFlags

* Fix clearTimeline at another spot.

Commit:
mark windows and macos chrome dev mode as flaky (#35495)

liyuqian committed:
Update macrobenchmarks README and app name (#35477)

engine-flutter-autoroll committed:
Roll engine ffba2f68d4e8..7d3e722349e7 (59 commits) (#35489)


https://github.com/flutter/engine/compare/ffba2f68d4e8...7d3e722349e7

git log ffba2f68d4e8ec2041f57226d907dcb929c8b795..7d3e722349e7923115f0b8800022a08c8dd795b3 --no-merges --oneline
7d3e72234 Roll buildroot to c5a493b25. (flutter/engine#9649)
8deeb77f0 make EmbeddedViewParams a unique ptr (flutter/engine#9640)
7862af546 Roll src/third_party/skia bd3d8d39b3e7..215ff3325230 (4 commits) (flutter/engine#9648)
a9ee687ac iOS PlatformView clip path (flutter/engine#9478)
c19b53cf2 Roll src/third_party/skia effee2065796..bd3d8d39b3e7 (1 commits) (flutter/engine#9647)
aeaa5edee Roll src/third_party/skia 2ef826576819..effee2065796 (1 commits) (flutter/engine#9646)
1cc1f04c9 Roll src/third_party/skia f2c52efce52b..2ef826576819 (2 commits) (flutter/engine#9645)
1c295b226 Roll src/third_party/skia 9fb7fa537d93..f2c52efce52b (1 commits) (flutter/engine#9644)
56a3f4102 Android Embedding Refactor PR31: Integrate platform views with the new embedding and the plugin shim. (flutter/engine#9206)
8ac7cbdda Fix warning about settings unavailable GN arg build_glfw_shell (flutter/engine#9642)
f66571781 Roll src/third_party/skia 21a486d04ae0..9fb7fa537d93 (21 commits) (flutter/engine#9639)
fd24007ca Roll Dart to 67ab3be10d35d994641da167cc806f20a7ffa679 (flutter/engine#9638)
628b174c9 Fixes a plugin overwrite bug in the plugin shim system. (flutter/engine#9589)
345d350ab Added Doxyfile. (flutter/engine#9632)
64b9eef5c Revert &#34;Roll Dart to 67ab3be10d35d994641da167cc806f20a7ffa679 (#9634)&#34; (flutter/engine#9637)
45e1ad274 Roll Dart to 67ab3be10d35d994641da167cc806f20a7ffa679 (flutter/engine#9634)
da6475eec Roll fuchsia/sdk/core/mac-amd64 from TRwIGIJLuznLNQzJk17zfboJrkErpe1XvGr0njCwemoC to byM-kyxL4bemlTYNqhKUfJfZoIUrCSzS6XzsFr4n9-MC (flutter/engine#9629)
bee12e6f3 Roll fuchsia/sdk/core/mac-amd64 from MBS_xxNZ_O32DxW1bhOeJisJqYG9JY_FAtSa3ZYupg4C to TRwIGIJLuznLNQzJk17zfboJrkErpe1XvGr0njCwemoC (flutter/engine#9626)
fea8fa6c4 Roll fuchsia/sdk/core/linux-amd64 from Efv1uHDvhLYgT8mvQmdAiJv7HiLix2L_kDRkK6P9ER4C to I2Qe1zxgckzIzMBTztvzeWYsDgcb9Fw-idSI16oIlx8C (flutter/engine#9625)
e7f8ca1de [all] add fuchsia.{net.NameLookup,posix.socket.Provider} (flutter/engine#9546)
a72a9997e Roll fuchsia/sdk/core/mac-amd64 from B-JzM_H7hG5kOIRFjmUqI3uWBAV7no8BnCTknbd2wQ4C to MBS_xxNZ_O32DxW1bhOeJisJqYG9JY_FAtSa3ZYupg4C (flutter/engine#9624)
b3fe3e95d Fix a race in the embedder accessibility unit test (flutter/engine#9585)
55818f602 Roll fuchsia/sdk/core/linux-amd64 from xtoj1ola0unTQOetly-V77CgpT6g8L1JUKWDqS8SuAQC to Efv1uHDvhLYgT8mvQmdAiJv7HiLix2L_kDRkK6P9ER4C (flutter/engine#9623)
37a54f2ab Roll fuchsia/sdk/core/linux-amd64 from O1niQGtIRghvjuMMCmxevRA1Y6seUn6onOao6Wii9hQC to xtoj1ola0unTQOetly-V77CgpT6g8L1JUKWDqS8SuAQC (flutter/engine#9622)
458a7647a Roll fuchsia/sdk/core/mac-amd64 from neDu8hWotIrKCQkxz1ScSJC4NoBva1_YVszr6wMbcZQC to B-JzM_H7hG5kOIRFjmUqI3uWBAV7no8BnCTknbd2wQ4C (flutter/engine#9621)
6e7b4ca44 Roll fuchsia/sdk/core/linux-amd64 from gMVpYn1cxQ0LeU-TSryUCg2o3rNcf7JWvlOqY6G00MYC to O1niQGtIRghvjuMMCmxevRA1Y6seUn6onOao6Wii9hQC (flutter/engine#9620)
e4d354d2c Roll src/third_party/skia 161f47dfbf6a..21a486d04ae0 (2 commits) (flutter/engine#9619)
28028759b Roll fuchsia/sdk/core/mac-amd64 from tPyg8rqV40gsoXEhDf7VskccnbJGCh4_bZp71YOUinEC to neDu8hWotIrKCQkxz1ScSJC4NoBva1_YVszr6wMbcZQC (flutter/engine#9618)
8e405c1db Roll fuchsia/sdk/core/mac-amd64 from UkCx2sMZsCM-w9nEuQC2TRfnJ7wjJCxsCxSDEx2uPegC to tPyg8rqV40gsoXEhDf7VskccnbJGCh4_bZp71YOUinEC (flutter/engine#9615)
609a98060 Fix uninitialized variables and put tests in flutter namespace. (flutter/engine#9613)
9807894ac Roll fuchsia/sdk/core/linux-amd64 from ur0ah3sh2atct83EqYX28SjG3fKt-7Driu48GbpdxmMC to gMVpYn1cxQ0LeU-TSryUCg2o3rNcf7JWvlOqY6G00MYC (flutter/engine#9612)
4e344e6cd Wire up custom event loop interop for the GLFW embedder. (flutter/engine#9089)
54c62263d Roll fuchsia/sdk/core/mac-amd64 from gyWAjP3BPfhpvHOOwaTusfA8JaGcY_UzjpoIGQnA_W0C to UkCx2sMZsCM-w9nEuQC2TRfnJ7wjJCxsCxSDEx2uPegC (flutter/engine#9611)
c3f8cab29 Roll fuchsia/sdk/core/linux-amd64 from xmxDtsnD0sfj7wxUaiMMhUwh72prBvMcYHY07lgTotcC to ur0ah3sh2atct83EqYX28SjG3fKt-7Driu48GbpdxmMC (flutter/engine#9610)
6f1a748ae Document various classes in //flutter/shell/common. (flutter/engine#9591)
4d3653092 Roll fuchsia/sdk/core/mac-amd64 from 4PD6FCl4NvKCavA0AVsdKtZPB3G5K72KprkEH0mr064C to gyWAjP3BPfhpvHOOwaTusfA8JaGcY_UzjpoIGQnA_W0C (flutter/engine#9609)
cf084bdb6 Roll fuchsia/sdk/core/linux-amd64 from XRYatTY5OvCnQ-5rGC8AnYltKa68CBxmnEK8QO0fpvQC to xmxDtsnD0sfj7wxUaiMMhUwh72prBvMcYHY07lgTotcC (flutter/engine#9607)
8d054008a disable mysterious failing tests (flutter/engine#9608)
aa817a9a2 Roll fuchsia/sdk/core/mac-amd64 from DeTFBSaxMBfZpfK0c7CifGpJbJLOrs3WtuCOaINwmrwC to 4PD6FCl4NvKCavA0AVsdKtZPB3G5K72KprkEH0mr064C (flutter/engine#9606)
4e988e0ae Roll fuchsia/sdk/core/linux-amd64 from jXpdljb7CHe8PEpUGGYqGvx6vFar6QRUh6HmpxMoS9sC to XRYatTY5OvCnQ-5rGC8AnYltKa68CBxmnEK8QO0fpvQC (flutter/engine#9605)
84ae36ac2 Roll fuchsia/sdk/core/mac-amd64 from IMr36r3rRLs1G7T5OtCudVMoWjPcjRbYGgyDg7LSxPwC to DeTFBSaxMBfZpfK0c7CifGpJbJLOrs3WtuCOaINwmrwC (flutter/engine#9604)
7135c8dbd Roll fuchsia/sdk/core/mac-amd64 from qcwCYvuT0PeU97HpvMVmC114EOKwfrk8PdFdZz_m6CIC to IMr36r3rRLs1G7T5OtCudVMoWjPcjRbYGgyDg7LSxPwC (flutter/engine#9602)
8f4df037b Roll fuchsia/sdk/core/linux-amd64 from Y3kUPtfq2frI60zx7VssO-WG733jtODCmnESyz0UGdEC to jXpdljb7CHe8PEpUGGYqGvx6vFar6QRUh6HmpxMoS9sC (flutter/engine#9601)
6c6a0d753 [trace clients] Remove fuchsia.tracelink.Registry (flutter/engine#9593)
f931539b2 Roll fuchsia/sdk/core/mac-amd64 from UsrGfX7Fj96MljgqUFc_A-o_ufsa_FX3eaG14ZSDWMAC to qcwCYvuT0PeU97HpvMVmC114EOKwfrk8PdFdZz_m6CIC (flutter/engine#9600)
...

Commit:
fix default artifacts to exclude ios and android (#35303)

tvolkert committed:
Mark update-packages as non-experimental (#35467)

Marking it as experimental was breaking tests and packaging
scripts on stable branches.

rami-a commented on packages/flutter/lib/src/material/material.dart:
```diff
@@ -217,9 +221,29 @@ class Material extends StatefulWidget {
+  ///   * [Material.elevation]
+  ///   * [ThemeData.brightness]
+  ///   * <https://material.io/design/color/dark-theme.html>
+  final bool darkThemeOverlay;
```
I think a verb in the bool name is preferred, something like `applyDarkThemeOverlay`

darrenaustin replied:
Good point. I was also thinking maybe it should include elevation as well. But then `applyDarkThemeElevationOverlay` is really getting to be a mouthful :).

rami-a commented on packages/flutter/lib/src/material/material.dart:
```diff
@@ -305,16 +330,42 @@ class _MaterialState extends State<Material> with TickerProviderStateMixin {
+        : elevation <= 12.0 ? 0.14
+        : elevation <= 16.0 ? 0.15
+        : 0.16;
+    final Color whiteOverlay = Color.fromRGBO(255, 255, 255, opacity);
```
This could be `Colors.white.withOpacity(opacity)`

darrenaustin replied:
Yeah, I thought of that, but didn't want to include 'colors.dart' just for `white`.  Looking at it again, that does look more readable.

rami-a reviewed (approved):
Looks pretty good but this will definitely have visually breaking changes for a lot of people. If anyone is using a dark theme they will have to then opt-out of this.

justinmc committed:
CupertinoTextField vertical alignment (#34723)

CupertinoTextField now supports vertical alignment via the textAlignVertical parameter.

goderbauer committed:
Re-apply 'Add currentSystemFrameTimeStamp to SchedulerBinding' (#35492)

goderbauer committed:
Include tags in SemanticsNode debug properties (#35491)

Commit:
Move usage flutter create tests into memory filesystem. (#35160)

chunhtai committed:
fix Selection handles position is off (#34665)

Commit:
add APK build time benchmarks (#35481)

dnfield committed:
Fix RenderFittedBox when child.size.isEmpty (#35487)

darrenaustin commented:
> Looks pretty good but this will definitely have visually breaking changes for a lot of people. If anyone is using a dark theme they will have to then opt-out of this.

Yeah, this concerns me as well. It does seem like it wouldn't be that drastic a visual change, as it only applies a slight white to elevated widgets in a dark theme. That said, I really hate breaking existing apps.

I have thought about making the default off and adding a property to `ThemeData` that would allow us to control it on at the theme level. That would allow a potential `ThemeData.from(colorScheme)` do the right thing here for new apps, but existing apps would be unchanged.

But that leads to more cruft and bloat in `ThemeData` for what may be a small breakage. 

Thoughts?

rami-a commented:
Hmm I actually do like the idea of adding this to `ThemeData`, despite how bloated it is becoming. 

Otherwise you would have to opt-out for each and every widget you create individually. So even if it is a big breaking change, switching one flag in your `ThemeData` to go back to what you had is a better experience.

I'm still on the fence though about whether it should be opt-in vs opt-out. But I'm definitely in favor of adding the flag to `ThemeData` to give broader control as well as being able to specify on a per-`Material` level.

srawlins committed:
Add type to StreamChannel in generated test code. (#35367)

This is necessary for those who wish to comply with static analysis like
`strict-raw-types: true`, which includes generated code within their package.

GaryQian committed:
Do not use ideographic baseline for RenderPargraph baseline (#35493)

HansMuller commented:
@rami-a I agree, it seems safest and least disruptive to make this an opt-in ThemeData flag.

darrenaustin committed:
Added applyDarkThemeElevationOverlay to ThemeData:

- Renamed Material.darkThemeOverlay -> Material.applyDarkThemeElevationOverlay
- Added applyDarkThemeElevationOverlay to ThemeData
- Modified Material to use the theme's overlay setting if none is provided
  to the widget.
- Turned the overlay off by default

darrenaustin commented:
I seemed to have borked this branch with some unfortunate git shenanigans. I have opened a new one that has a clean history: #35560.
